reactivesocket-client/src/main/java/io/reactivesocket/client/LoadBalancer.java

```diff
@@ -121,18 +174,43 @@ public LoadBalancer(Publisher<List<ReactiveSocketFactory<SocketAddress>>> factor
     }
 
     private synchronized void addSockets(int numberOfNewSocket) {
-        activeFactories.entrySet()
-            .stream()
-            // available factories that don't map to an already established socket
-            .filter(e -> !activeSockets.containsKey(e.getKey()))
-            .map(e -> e.getValue())
-            .filter(factory -> factory.availability() > 0.0)
-            .sorted((a, b) -> -Double.compare(a.availability(), b.availability()))
-            .limit(numberOfNewSocket)
-            .forEach(factory -> {
-                pendingSockets += 1;
-                factory.apply().subscribe(new SocketAdder(factory.remote()));
-            });
+        int n = Math.min(activeFactories.size(), numberOfNewSocket);
+        int maxEffort = 1 + activeFactories.size()/2;
+        int effort = 0;
+        while (n > 0 && effort < maxEffort) {
+            ReactiveSocketFactory<?> factory0 = activeFactories.poll();
+            ReactiveSocketFactory<?> factory1 = activeFactories.poll();
+            if (factory0 == null) {
+                // queue is empty
+                break;
+            }
+            if (factory1 == null) {
+                // factory0 was the last item in the queue
+                if (factory0.availability() > 0.0) {
+                    pendingSockets++;
+                    factory0.apply().subscribe(new SocketAdder(factory0));
+                } else {
+                    activeFactories.addLast(factory0);
+                }
+                break;
+            }
+
+            if (factory0.availability() == 0.0 && factory1.availability() == 0.0) {
+                activeFactories.addLast(factory0);
+                activeFactories.addLast(factory1);
+                effort++;
+            } else if (factory0.availability() < factory1.availability()) {
+                n--;
+                pendingSockets++;
+                activeFactories.addLast(factory0);
```

Comment:
Basically, what I'm trying to achieve here is selecting a good factory without having to scan the entire list.
If I don't care about having a "good" factory, I can just select the top of the queue, but, by looking at the first 2 elements it gives me almost enough information for making a smart decision (this isn't exactly what the power of two choices paper proved but the idea is similar).

The reason why I'm "aging" (adding to the end) the unselected factory, is to avoid having a bad instance in the front, which will always lead to selecting the second one.

The other option here would be to use literally the Power of two choices algorithm (pick two factories randomly, not only the first two), which is something I'm ok to consider. It would have roughly the same characteristic. 

If you think it's simpler to understand, I'll make the change.